FreeBSD initial support

rayrapetyan opened the pull request:
This pull request adds initial support for FreeBSD.
It's primarily based on fbsd branch from: github.com:asomers/delve.git, just fixes few critical bugs and adds few new functions.

It passes almost all tests (281 of 288), except few "Concurrency" ones, but I believe it's not a bug, but a natural behavior for FreeBSD and these tests should be slightly changed.

E.g. on "TestBreakpointCounts" two threads are expected to hit same breakpoint 100 times each. In FreeBSD, when main thread exits, all "child" threads are not triggering TRAP signals anymore, process just finishes. That's why when main thread reaches 100 breakpoint hits and exits, child thread will never reach 100 and stops somewhere around 95. Sometimes, when child thread is faster and completes BP race first, we can see this test as PASSED. 

Other than that it works fine.

ref: https://github.com/go-delve/delve/issues/213

Thanks.

aarzilli commented:
In bpcountstest.go (the fixture used by TestBreakpointCounts) the main goroutine specifically waits for the two child goroutines before exiting. If TestBreakpointCounts doesn't pass either there's a bug in your backend implementation or in Go. I think it's more likely that there's a bug here.

TestBreakpointCounts actually checks that the backend correctly detects when the same breakpoint is hit simultaneously on two different threads. Not doing this correctly is a common mistake, it was made by different people on the darwin, linux and windows native backend. It's likely that you have the same problem.

aarzilli commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,368 @@
+
+// Used by ContinueOnce
+// stop stops all running threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
```
How do threads other than the trapthread get stopped? Since you aren't callling Thread.stop here.

rayrapetyan replied:
In FreeBSD, whenever one thread is trapped, the entire process (all threads) are stopped. Since we call "stop" only after "trapWait" (in ContinueOnce), all threads are always stopped, so no need to explicitly check for that.

aarzilli commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,368 @@
+// OSProcessDetails contains FreeBSD specific
+// process details.
+type OSProcessDetails struct {
+	comm string
```
I can't find where this is used.

rayrapetyan replied:
Seems it's not used anywhere. I've tried to use tid (main process thread id) in singleStep mimicking comparison in line 59 from linux_threads.go:
`if (status == nil || status.Exited()) && wpid == t.dbp.pid`
But later replaced all logic in singleStep with a single call to trapWait(). Btw can we do the same for Linux? The problem with Linux's singleStep implementation is that it doesn't handle threads add\remove properly. I see that we are waiting for a TRAP from "this" thread ignoring signals from other threads. Maybe we can just pass tid into trapWait as additional param to simulate the same?

aarzilli replied:
> Btw can we do the same for Linux?

see the comment in (*Process).wait in proc_linux.go

rayrapetyan replied:
The only comment there is about a workaround for lock when main thread exits and leaves zombies. Could you clarify how does that rely to replacing logic in singleStep() with a call to waitTrap()?

aarzilli replied:
The full wait is slow because of the workaround. Also all threads are stopped and we are only executing one instruction of threads stopped on a breakpoint, so it's fine.

rayrapetyan replied:
On FreeBSD, even when all threads are stopped, wait in cycle may return multiple queued signals from multiple stopped threads.
 
I'm preparing a detailed step by step explanation of why "TestBreakpointCounts" fails, hopefully someone can help with a fix.

aarzilli commented on pkg/proc/native/proc_freebsd.c:
```diff
@@ -0,0 +1,93 @@
+	return (command_name);
+}
+
+int find_status(int pid){
```
Looking at the [procfs manpage](https://www.freebsd.org/cgi/man.cgi?query=procfs&apropos=0&sektion=0&manpath=FreeBSD+10.4-RELEASE&arch=default&format=html) it seems to me that this and find_executable could be implemented using that. Am I wrong?

rayrapetyan replied:
It is not recommended to use procfs in FreeBSD as it requires some additional explicit changes in /etc/fstab (https://www.freebsd.org/doc/en/articles/linux-users/procfs.html), not all users have this Linux-compatibility option enabled. Using procstat calls is a recommended way to get info.

aarzilli replied:
Is it possible to do this without cgo?

rayrapetyan replied:
I believe no (except procfs which requires additional explicit mount).

rayrapetyan commented:
So far I was able to reproduce "TRAP events disappear" issue with a pure C code, without go stuff involved. Under some circumstances, app is exiting without delivering queued TRAP events to debugger. Preparing a message to freebsd-hackers group.

rayrapetyan commented:
FreeBSD bug created: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=236035

aarzilli commented:
Thank you, let me know how it goes.

derekparker commented:
Finished a cursory review this morning, will be doing a more in depth review tomorrow or the following day. In the meantime, I saw some commented out code, I can just add a blanket statement now to clean all that up.

By the way, thanks for submitting this patch!

aarzilli commented:
Regarding the discussion around TestBreakpointCounts I'm ok with disabling that test on freebsd and merging this now to prevent this PR from coderotting. There's also a free CI service that has FreeBSD servers: https://cirrus-ci.org/guide/FreeBSD/.

rayrapetyan commented on pkg/proc/native/threads_freebsd.go:
```diff
@@ -0,0 +1,144 @@
+		}
+
+		t.dbp.trapWait(t.dbp.pid)
+		/*_, status, err := t.dbp.waitFast(t.dbp.pid)
```
The problem here is that this logic seems to be broken in any existing implementation, e.g. in Linux waitFast is used, then exiting conditions checked, but there is no proper handling of thread creation\death in this place. trapWait handles all these cases properly.

derekparker replied:
Ack, understood and makes sense.

rayrapetyan commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,368 @@
+	}
+}
+
+func (dbp *Process) loadProcessInformation() {
```
Removed from here and Linux implementation.

rayrapetyan commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,368 @@
+
+// Used by Detach
+func (dbp *Process) detach(kill bool) error {
+	err := PtraceDetach(dbp.pid, 0)
```
Fixed.

rayrapetyan commented on pkg/proc/native/ptrace_freebsd.go:
```diff
@@ -0,0 +1,112 @@
+	var info C.struct_ptrace_lwpinfo
+	_, err = C.ptrace_lwp_info(C.int(wpid), &info)
+
+	/*
```
Removed.

rayrapetyan commented on pkg/proc/native/threads_freebsd.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+func (t *Thread) Blocked() bool {
+	regs, err := t.Registers(false)
```
Changed.

rayrapetyan commented on pkg/proc/native/threads_freebsd.go:
```diff
@@ -0,0 +1,144 @@
+			return err
+		}
+
+		t.dbp.trapWait(t.dbp.pid)
```
Fixed.

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,358 @@
+		return errors.New("could not deliver signal " + err.Error())
+	}
+	// If the process is stopped, we must continue it so it can receive the signal
+	PtraceCont(dbp.pid, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,358 @@
+
+// Used by initialize
+func (dbp *Process) updateThreadList() error {
+	tids := PtraceGetLwpList(dbp.pid)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,358 @@
+		}
+	}
+	// all threads are resumed
+	PtraceCont(dbp.pid, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/threads_freebsd.go:
```diff
@@ -0,0 +1,128 @@
+	}
+	// If the process is stopped, we must continue it so it can receive the
+	// signal
+	PtraceCont(t.dbp.pid, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/threads_freebsd.go:
```diff
@@ -0,0 +1,128 @@
+		if th.ID == t.ID {
+			break
+		}
+		PtraceCont(th.ID, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,358 @@
+		if status.StopSignal() == sys.SIGTRAP {
+			if pl_flags&sys.PL_FLAG_EXITED != 0 {
+				delete(dbp.threads, tid)
+				PtraceCont(tid, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,358 @@
+
+// Used by Detach
+func (dbp *Process) detach(kill bool) error {
+	return PtraceDetach(dbp.pid, 0)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

rayrapetyan replied:
Doesn't work that way.

derekparker commented on pkg/proc/native/registers_freebsd_amd64.go:
```diff
@@ -0,0 +1,92 @@
+		return nil, err
+	}
+	var fsbase int64
+	err = sys.PtraceGetFsBase(thread.ID, &fsbase)
```
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

rayrapetyan commented:
@derekparker, wrapped all places except PtraceDetach - it doesn't work that way. Thanks.

derekparker commented:
@rayrapetyan thanks for the prompt changes! The last thing I have is to disable any concurrent tests that are failing due to the FreeBSD issue with an informative `t.Skip(...)` and then we should be able to merge this and get FreeBSD CI set up.

rayrapetyan commented:
@derekparker, there are few weird test cases. When running: 
go test
from pkg/proc folder, two tests always fail:
```
--- FAIL: TestAttachDetach (0.08s)
    proc_test.go:136: Did not continue to correct location, expected line :11 got /usr/local/go/src/runtime/panic.go:681
        	at /ara/devel/blockchain/go/src/github.com/go-delve/delve/pkg/proc/proc_test.go:2887
--- FAIL: TestIssue844 (0.14s)
    proc_test.go:113: failed assertion at proc_test.go:3204: Continue - ptraceGetLwpInfo err no such process -1
i: 0 -> equalsTwo: false 
i: 1 -> equalsTwo: false 
0
fatal error: all goroutines are asleep - deadlock!
```
 
But whenever you run each of these two individually (e.g. go test -run TestAttachDetach) - they always PASS. Any thoughts of why "go test" may perform like that? Thanks.

aarzilli commented:
The TestAttachDetach could happen if you are running tests in parallel (like `go test ./...`), the other one I don't know, it needs to be debugged.

rayrapetyan commented:
I've tried `cd pkg/proc; go test -parallel 1` - it also fails...

aarzilli commented:
It could be that Detach(true) isn't actually killing the target process, one instance of testnextnethttp stays running and makes TestAttachDetach fail.

rayrapetyan commented:
"Skipped" concurrent tests. Please let me know if I can help to setup FreeBSD CI in Travis.
@aarzilli, sorry, just found your "go mod" related question. Please let me know how to perform vendor code update procedure properly and I'll commit missed fixes. Thanks.

derekparker commented:
@rayrapetyan thanks! You should just have to run `go mod vendor`.

rayrapetyan commented:
@derekparker, hm, `go mod vendor` just re-creates original vendor directory vanishing all my changes. Does that mean I should commit everything into original repo first (e.g. into golang.org/x/sys)? Thanks!

derekparker commented:
@rayrapetyan are you making manual changes to the vendor directory? Yeah, that won't work for the reason you described.

If you've made any changes to any dependency of Delve, you must commit that change upstream and then re-vendor it in Delve (from upstream). I didn't take a close look at what changes were made in the vendor dir, but is it possible to implement without any code changes to our dependencies?

rayrapetyan commented:
@derekparker, @aarzilli, I've cleared vendor-related changes, the only artifacts left are constants for ptrace and registers-related structs, I believe they should stay in vendors. Please suggest how to proceed. I've never committed into golang core, it requires a special procedure (https://golang.org/doc/contribute.html), so if you have accounts there could you commit these changes? Thanks.

aarzilli commented:
I'm taking a look at the tests you have disabled. It's quite a lot of tests, if they all need to be disabled it seems that there is something that's pretty thoroughly broken here.

As for the changes to golang.org/x/sys/unix, I don't have a freebsd install and I don't know pretty much anything about it, I can't make them for you. 

For the constants it's easier to copy them to our side. The changes to Stat_t, Dirent, etc are worrying however, that's an ABI change, I don't know that's handled on golang.org/x/sys/unix.

It's not that hard, you just have to follow steps 0 through 4 at https://golang.org/doc/contribute.html. Ignore the rest of that page, then you make a branch for your changes and mail them with `git codereview mail`.

rayrapetyan commented:
@aarzilli, these are all parallel_next and concurrentstep tests, there are also two other tests I will investigate. Will also proceed with commits into sys/unix, thanks!

aarzilli commented:
> these are all parallel_next and concurrentstep tests, there are also two other tests I will investigate

My guess is that when a thread executes an INT 3 freebsd will send us immediately a signal: if other threads are running on different cores they will keep running until their quantum exipres, but the signal will be sent immediately. Which means that when (*Process).stop is called some of the threads could still be running and that's why we are missing some breakpoints in all those tests.

rayrapetyan commented:
golang.org: commit for ptrace support on FreeBSD: https://go-review.googlesource.com/c/go/+/166423

derekparker commented:
@rayrapetyan what's the status of the above patch? Looks like there might be some outstanding review comments to address before it's considered for merging.

rayrapetyan commented:
@derekparker, I was stuck with committing changes into golang - along with i386 changes which I provided they require changes for CPU architectures I don't own (ARM) and therefore can't test...

derekparker commented:
@rayrapetyan It looks like all you need to do is generate the files for ARM. I think if you do that and submit it to the patch, the burden for review and testing will be on them. They can use their own internal builders for testing.

rayrapetyan commented:
Pushed "generated" ARM files. Awaiting merge.
https://github.com/golang/go/pull/30704

derekparker commented:
@rayrapetyan looks like that PR was closed because that stdlib package is locked, shoulda caught that earlier. In any event, can you open a PR against `x/sys/unix` if you haven't already? We are already vendoring that so it would be easier to vendor your changes once merged.

rayrapetyan commented:
@derekparker, awaiting approval on https://github.com/golang/sys/pull/34.

derekparker commented:
@rayrapetyan ack, thanks.

derekparker commented:
@rayrapetyan looks like it's been stagnant for a while. I just pinged some folks to get that pushed through soon hopefully.

rayrapetyan commented:
Yeah, thanks!

On Tue, Jun 18, 2019 at 1:21 PM Derek Parker <notifications@github.com>
wrote:

> @rayrapetyan <https://github.com/rayrapetyan> looks like it's been
> stagnant for a while. I just pinged some folks to get that pushed through
> soon hopefully.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-delve/delve/pull/1480?email_source=notifications&email_token=ABD3S26PUZKHH5O3DFKUXI3P3FNWNA5CNFSM4GXSZ7HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYAEGTQ#issuecomment-503333710>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABD3S22UVDCNUVIIIUO6IHLP3FNWNANCNFSM4GXSZ7HA>
> .
>

derekparker commented:
@rayrapetyan looks like it was merged! Can you update this PR by vendoring these new changes please.

rayrapetyan commented:
Done!
Wondering why do we need to carry “vendor” folder in git repo? Can’t we
just make ‘go mod vendor’ an installation step? Thanks!

On Mon, Jun 24, 2019 at 8:00 AM Derek Parker <notifications@github.com>
wrote:

> @rayrapetyan <https://github.com/rayrapetyan> looks like it was merged!
> Can you update this PR by vendoring these new changes please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-delve/delve/pull/1480?email_source=notifications&email_token=ABD3S22LJCBOSSO7ESEDQ7TP4DOQBA5CNFSM4GXSZ7HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYNG32A#issuecomment-505048552>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABD3S2YVPRKT47LTTLHTQ23P4DOQBANCNFSM4GXSZ7HA>
> .
>

aarzilli commented:
>Wondering why do we need to carry “vendor” folder in git repo?

the vendor folder makes delve compile in GOPATH mode and on pre-modules version of go. It also makes this project self-contained, preventing left-pad style accidents.

derekparker commented:
@rayrapetyan looks like the last update broke everything lol. How did you vendor it?

rayrapetyan commented:
Oh... weird. I did "go mod vendor", it updated go.mod and go.sum and vendor folder, I've committed these changes, that's it. What had I missed?
Now it fails in sys\windows (some missed definitions which I can see defined in sources), linux: x/crypto (misses definitions which are there, x/crypto separately passes all tests) and something in Mac...

derekparker commented:
@rayrapetyan does it work when you build and test locally?

rayrapetyan commented:
Yes, it passes all tests locally in FreeBSD. I can reproduce issue with x/crypto in Ubuntu VM. Do you know how to make x/crypto build correctly?

derekparker commented:
@rayrapetyan the vendor has been updated with your changes via #1600 thanks to @aarzilli. You should be able to just remove your last commit and rebase on master.

rayrapetyan commented:
@derekparker, @aarzilli, thanks guys, I've spent several hours trying to make latest x/crypto to work with x/sys with no luck. Hopefully we can merge my branch now.

derekparker commented on pkg/proc/fbsdutil/regs.go:
```diff
@@ -0,0 +1,416 @@
+// Get returns the value of the n-th register (in x86asm order).
+func (r *AMD64Registers) Get(n int) (uint64, error) {
+	reg := x86asm.Reg(n)
+	const (
```
These bitmasks are incorrect, see: https://github.com/go-delve/delve/pull/1583.

derekparker commented on pkg/proc/fbsdutil/regs.go:
```diff
@@ -0,0 +1,416 @@
+	return
+}
+
+const (
```
Instead of duplicating this, let's reuse what already exists in `linutil`.

derekparker commented on pkg/proc/gdbserial/gdbserver.go:
```diff
@@ -1327,6 +1327,9 @@ func (p *Process) loadGInstr() []byte {
 	case "darwin":
 		// mov rcx,QWORD PTR gs:{uint32(off)}
 		op = []byte{0x65, 0x48, 0x8B, 0x0C, 0x25}
+	case "freebsd":
```
Let's clean this up a bit by just merging this into the case above, e.g. `case "darwin", "freebsd":`

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,367 @@
+	var err error
+	dbp.execPtraceFunc(func() { err = sys.PtraceLwpEvents(dbp.pid, 1) })
+	if err == syscall.ESRCH {
+		// XXX why do we wait here?
```
Is this still an open question?

rayrapetyan commented:
@derekparker, updated. An interesting finding - with fixed bitmasks another concurrent test is constantly failing, so I've added it to skipped for now. Definitely bitmasks affect other concurrent tests also. I will look into all skipped tests after July 20.

derekparker commented:
@rayrapetyan hmm, that's odd. That change didn't break any tests on any currently supported OS.

I have one last request before merging. The `t.Skip(...)` message isn't exactly correct. The tests are valid, just broken on the OS. Please reword and link to relevant issues (like the upstream freebsd issue you created) in a comment. After that I will merge promptly and create some follow up issues to track fixing the broken tests.

derekparker commented:
@aarzilli I'm planning on merging this by EOD today even if the above comment isn't addressed (and just do my own follow up commit if need be). Is there any concern you have with this PR or are you good with it at this point? 

I have questions on why some of the tests are failing and we need to chase those down (@rayrapetyan has mentioned starting after July 20). I'm eager to get this merged, but as we are also _really_ close to a release and I want to make sure we're not releasing something big like a new OS without very much vetting. The upside of merging soon however is we can start getting early adopters to test and provide feedback.

derekparker commented on pkg/proc/native/proc_freebsd.go:
```diff
@@ -0,0 +1,366 @@
+		if err != nil {
+			return nil, fmt.Errorf("wait err %s %d", err, pid)
+		}
+		if status.Killed() {
```
I wonder if this has anything to do with the failing concurrent tests, since we're effectively ignoring `status.Killed`, yes we send `SIGKILL` to the process in `process.kill()`. I have a freebsd VM I can test this theory out on tomorrow.

derekparker commented:
@rayrapetyan a huge thanks to you for this port! Now looking forward to getting the concurrency issues sorted out.
